### PR TITLE
[FIX] microsoft_calendar : Apply improvement on 14.0

### PR DESCRIPTION
**Pull request by fdamhaut**
Apply improvement of microsoft calendar on 14.0 to fix different reported issues.

For more information see #68700

opw-2658427

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by remybar** [comment]
> @arj-odoo @remybar Un de vous deux peux jeter un oeil ? :D

Could you:
- add a real commit message to explain the issues fixed by your PR (it will be far easier for us to review your code then),
- give us some information about the tests you did to validate this PR (it will be easier for us to think about some missing corner cases for example).

Thanks!

**Comment by fdamhaut** [comment]
> Could you:
> 
> * add a real commit message to explain the issues fixed by your PR (it will be far easier for us to review your code then),
> * give us some information about the tests you did to validate this PR (it will be easier for us to think about some missing corner cases for example).
> 
> Thanks!

I try to improve the commit message but the commit is really just as i mentioned a porting of the possible feature that @arj-odoo implemented on master with the restriction on what can be changed in stable

On top of that, I also tested to see if current DB don't raise error when being used with the new code and that doesn't seem to create issues.

On the other hand, I did not change the test compared to the one that were written in master and ported them because it was an easy and useful addition.

TL;DR this is a back-port of the change in calendar to solve bug and all the different code change were already documented by ARJ in the PR for master that is linked

**Review by arj-odoo — CHANGES_REQUESTED** [review]
Hello, This PR is quite huge for a bugfix. I understand that you cherry-picked the improvements in master for google but I think the changes are quite intrusive. In a bugfix, you need to change the minimum to make it work. I lack time to test it deepfully (reproduce the exact issue of the ticket in a local DB, apply the changes of this PR and see if they are gone). 

_get_microsoft_sync_domain modifications are OK. dont_modify context key should be tested to see if they are relevant everywhere. 

Code in master have been tested broadly but @remybar found issues. Please, avoid to backport known issues. It is not always possible to cleanly backport fixes. 
These changes MUST be tested by a PO or Andree (avg) on a runbot before going forward.

**Comment by avg-odoo** [comment]
@fdamhaut Hello - since the build of https://runbot.odoo.com/runbot/bundle/14-0-opw-2658427-flda-93766 dropped, would you please confirm that we can test on https://10618864-14-0-test_lint.runbot42.odoo.com/ ?
cc @kbr-odoo

File: addons/microsoft_calendar/utils/microsoft_calendar.py [hunk 1]
```diff
@@ -88,6 +91,13 @@ def delete(self, event_id, token=None, timeout=TIMEOUT):
                 raise e
             _logger.info("Microsoft event %s was already deleted" % event_id)
 
+    @requires_auth_token
+    def answer(self, event_id, answer, values, token=None, timeout=TIMEOUT):
+        url = "/v1.0/me/calendar/events/%s/%s" % (event_id, answer)
+        headers = {'Content-type': 'application/json', 'Authorization': 'Bearer %s' % token}
+        self.microsoft_service._do_request(url, json.dumps(values), headers, method='POST', timeout=timeout)
```
Comment (remybar) [inline addons/microsoft_calendar/utils/microsoft_calendar.py:98]: No need to use `json.dumps(values, separators=(',', ':'))` instead of `json.dumps(values)` here ?

File: addons/microsoft_calendar/models/calendar.py [hunk 2]
```diff
@@ -54,10 +56,26 @@ def write(self, values):
             if previous_event_before_write != previous_event_after_write:
                 # Outlook returns a 400 error if you try to synchronize an occurrence of this type.
                 raise UserError(_("Modified occurrence is crossing or overlapping adjacent occurrence."))
-        return super().write(values)
+        notify_context = self.env.context.get('dont_notify', False)
+        res = super(Meeting, self.with_context(dont_notify=notify_context)).write(values)
+        if recurrence_update_setting in ('all_events',) and len(self) == 1 and values.keys() & self._get_microsoft_synced_fields():
```
Comment (arj-odoo) [inline addons/microsoft_calendar/models/calendar.py:61]: recurrence_update_setting in ('all_events',) can't be in all_events because updating the whole recurrence was introduced in 15.0.

File: addons/microsoft_calendar/models/calendar.py [hunk 3]
```diff
@@ -81,18 +99,25 @@ def _microsoft_to_odoo_values(self, microsoft_event, default_reminders=(), defau
...
             'show_as': 'free' if microsoft_event.showAs == 'free' else 'busy',
             'recurrency': microsoft_event.is_recurrent()
         }
+        if commands_partner:
+            # Add partner_commands only if set from Microsoft. The write method on calendar_events will
+            # override attendee commands if the partner_ids command is set but empty.
+            values['partner_ids'] = commands_partner
+
+        if microsoft_event.is_recurrent() and not microsoft_event.is_recurrence():
+            # Propagate the follow_recurrence according to the google result
```
Comment (arj-odoo) [inline addons/microsoft_calendar/models/calendar.py:119]:
```suggestion
            # Propagate the follow_recurrence according to the microsoft result
```

File: addons/microsoft_calendar/tests/test_sync_microsoft2odoo.py [hunk 4]
```diff
@@ -274,4 +274,4 @@ def test_cancel_occurence_of_recurrent_event(self):
         self.assertEqual(recurrent_event.calendar_event_ids[0].start, datetime(2021, 7, 15, 15, 00))
         self.assertEqual(recurrent_event.calendar_event_ids[0].stop, datetime(2021, 7, 15, 15, 30))
         self.assertEqual(recurrent_event.calendar_event_ids[1].start, datetime(2021, 7, 17, 15, 00))
-        self.assertEqual(recurrent_event.calendar_event_ids[1].stop, datetime(2021, 7, 17, 15, 30))
+        self.assertEqual(recurrent_event.calendar_event_ids[1].stop, datetime(2021, 7, 17, 15, 30))
```
Comment (remybar) [inline addons/microsoft_calendar/tests/test_sync_microsoft2odoo.py:277]: a newline is missing :nerd_face:

File: addons/microsoft_calendar/models/calendar_recurrence_rule.py [hunk 5]
```diff
@@ -89,30 +132,22 @@ def _microsoft_to_odoo_values(self, microsoft_recurrence, default_reminders=(),
         }
 
     def _microsoft_values(self, fields_to_sync):
-        events = self.calendar_event_ids.sorted('start')
         events_outliers = self.calendar_event_ids.filtered(lambda e: not e.follow_recurrence)
+        events = self.calendar_event_ids.sorted('start')
```
Comment (arj-odoo) [inline addons/microsoft_calendar/models/calendar_recurrence_rule.py:136]: why moving this line below events_outliers ?
useless diff

File: addons/microsoft_calendar/models/microsoft_sync.py [hunk 6]
```diff
@@ -174,15 +170,18 @@ def _sync_recurrence_microsoft2odoo(self, microsoft_events: MicrosoftEvent):
                 else:
                     value = self.env['calendar.event']._microsoft_to_odoo_values(recurrent_event, (), default_values)
 
+
```
Comment (arj-odoo) [inline addons/microsoft_calendar/models/microsoft_sync.py:173]: useless diff

File: addons/microsoft_calendar/tests/test_sync_microsoft2odoo.py (hunk 4 above)
Comment (arj-odoo) [inline addons/microsoft_calendar/tests/test_sync_microsoft2odoo.py:277]: missing new line

**Review by arj-odoo — COMMENTED** [review]
Just a few cosmetic changes before I checkout your branch :-).
Nothing crucial at this point :-).

File: addons/microsoft_calendar/utils/microsoft_calendar.py [hunk 7]
```diff
@@ -66,14 +69,14 @@ def insert(self, values, token=None, timeout=TIMEOUT):
         headers = {'Content-type': 'application/json', 'Authorization': 'Bearer %s' % token}
         if not values.get('id'):
             values.pop('id', None)
-        status, data, time = self.microsoft_service._do_request(url, json.dumps(values), headers, method='POST', timeout=timeout)
+        dummy, data, dummy = self.microsoft_service._do_request(url, json.dumps(values, separators=(',', ':')), headers, method='POST', timeout=timeout)
```
[the remark this answers is not in the document]
Comment (arj-odoo) [reply addons/microsoft_calendar/utils/microsoft_calendar.py:72]: > use `_` instead of `dummy`, so `_, data, _`.

Hum no, because it could override the _ import used for the translations. dummy is fine.

File: addons/microsoft_calendar/utils/microsoft_calendar.py (hunk 1 above)
Comment (arj-odoo) [reply addons/microsoft_calendar/utils/microsoft_calendar.py:98]: Indeed I added the separators in google but it is not helpful as it is the default value.

File: addons/microsoft_calendar/utils/microsoft_calendar.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (remybar) [reply addons/microsoft_calendar/utils/microsoft_calendar.py:72]: Ah yes ... I didn't notice that `_` was imported from odoo :P

File: addons/microsoft_calendar/models/calendar.py [hunk 8]
```diff
@@ -81,18 +99,25 @@ def _microsoft_to_odoo_values(self, microsoft_event, default_reminders=(), defau
         values = {
             **default_values,
             'name': microsoft_event.subject or _("(No title)"),
-            'description': microsoft_event.bodyPreview,
+            'description': plaintext2html(microsoft_event.bodyPreview),
```
Comment (remybar) [inline addons/microsoft_calendar/models/calendar.py:102]: Do not use `plaintext2html` as in `14.0`, the `description` field is a `field.Text` and not yet a `field.Html` ;-)

File: addons/microsoft_calendar/models/calendar_recurrence_rule.py [hunk 9]
```diff
@@ -19,8 +16,19 @@ class RecurrenceRule(models.Model):
...
+        for recurrence in self:
+            if recurrence.rrule != recurrence._rrule_serialize():
+                recurrence.write({'rrule': recurrence._rrule_serialize()})
+
+    def _inverse_rrule(self):
+        for recurrence in self.filtered('rrule'):
+            values = self._rrule_parse(recurrence.rrule, recurrence.dtstart)
+            # We avoid to trigger patch in the inverse
+            recurrence.write(values)
+
```
Comment (arj-odoo) [inline addons/microsoft_calendar/models/calendar_recurrence_rule.py:29]: I think you can remove this inverse, it does nothing different than the method in calendar_recurrence.py file

**Comment by arj-odoo** [comment]
@robodoo r+
